internal/project/custom/custom.go

```diff
@@ -137,26 +145,80 @@ func (step *Step) CompileDrone(output *drone.Output) error {
 		return true
 	}
 
-	droneStep := drone.MakeStep(step.Name()).
-		DependsOn(dag.GatherMatchingInputNames(step, droneMatches)...)
+	baseDroneStep := func() *drone.Step {
+		droneStep := drone.MakeStep(step.Name())
 
-	if step.Drone.Privileged {
-		droneStep.Privileged()
-	}
+		if step.Drone.Privileged {
+			droneStep.Privileged()
+		}
 
-	if step.Drone.Requests != nil {
-		droneStep.ResourceRequests(step.Drone.Requests.CPUCores, step.Drone.Requests.MemoryGiB)
-	}
+		if step.Drone.Requests != nil {
+			droneStep.ResourceRequests(step.Drone.Requests.CPUCores, step.Drone.Requests.MemoryGiB)
+		}
 
-	for k, v := range step.Drone.Environment {
-		droneStep.Environment(k, v)
+		for k, v := range step.Drone.Environment {
+			droneStep.Environment(k, v)
+		}
+
+		for _, volume := range step.Drone.Volumes {
+			droneStep.EmptyDirVolume(volume.Name, volume.MountPath)
+		}
+
+		return droneStep
 	}
 
-	for _, volume := range step.Drone.Volumes {
-		droneStep.EmptyDirVolume(volume.Name, volume.MountPath)
+	output.Step(baseDroneStep().
+		DependsOn(dag.GatherMatchingInputNames(step, droneMatches)...))
+
+	for _, customPipeline := range step.Drone.Pipelines {
+		pipeline := output.Pipeline(customPipeline.Name, customPipeline.Triggers, customPipeline.Crons)
+
+		type baseDroneStepper interface {
+			BuildBaseDroneSteps(output drone.StepService)
+		}
+
+		baseStepNames := []string{}
+
+		for _, baseInput := range dag.GatherMatchingInputsRecursive(step, dag.Implements[baseDroneStepper]()) {
+			baseInput.(baseDroneStepper).BuildBaseDroneSteps(pipeline) //nolint:forcetypeassert // type is checked in GatherMatchingInputsRecursive
+
+			baseStepNames = append(baseStepNames, baseInput.Name())
+		}
+
+		pipeline.Step(
+			drone.CustomStep("load-artifacts",
+				"s3cmd --host=rook-ceph-rgw-ci-store.rook-ceph.svc --host-bucket=rook-ceph-rgw-ci-store.rook-ceph.svc --no-ssl --stats sync s3://${CI_COMMIT_SHA}${DRONE_TAG//./-} .",
+			).
+				DependsOn(baseStepNames...).
+				EnvironmentFromSecret("AWS_ACCESS_KEY_ID", "rook_access_key_id").
+				EnvironmentFromSecret("AWS_SECRET_ACCESS_KEY", "rook_secret_access_key"),
+		)
+
+		droneStep := baseDroneStep()
+
+		for k, v := range customPipeline.EnvironmentOverride {
+			droneStep.Environment(k, v)
+		}
+
+		pipeline.Step(droneStep.DependsOn("load-artifacts"))
 	}
 
-	output.Step(droneStep)
+	if len(step.Drone.Pipelines) > 0 {
+		// add a "save artifacts" step to the default pipeline
+		output.Step(
```

Comment:
do we always need it? maybe another boolean in the yaml to enabled load/save :thinking:

Reply:
it's only if the pipelines are enabled

Reply:
but I think without load/save it's quite useless